group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -0,0 +1,959 @@
+     * when the operation is completed or an exception if the write operation failed.
+     *
+     * @param <T> The type of the result.
+     */
+    public <T> CompletableFuture<T> scheduleWriteOperation(
+        String name,
+        TopicPartition tp,
+        CoordinatorWriteOperation<S, T, U> op
+    ) {
+        log.debug("Scheduled execution of write operation {}.", name);
+        CoordinatorWriteEvent<T> event = new CoordinatorWriteEvent<>(name, tp, op);
+        processor.enqueue(event);
+        return event.future;
+    }
+
+    /**
+     * Schedules a read operation.
+     *
+     * @param name  The name of the write operation.
+     * @param tp    The address of the coordinator (aka its topic-partitions).
+     * @param op    The read operation.
+     *
+     * @return A future that will be completed with the result of the read operation
+     * when the operation is completed or an exception if the write operation failed.
+     *
+     * @param <T> The type of the result.
+     */
+    public <T> CompletableFuture<T> scheduleReadOperation(
+        String name,
+        TopicPartition tp,
+        CoordinatorReadOperation<S, T> op
+    ) {
+        log.debug("Scheduled execution of read operation {}.", name);
+        CoordinatorReadEvent<T> event = new CoordinatorReadEvent<>(name, tp, op);
+        processor.enqueue(event);
+        return event.future;
+    }
+
+    /**
+     * Schedules an internal event.
+     *
+     * @param name  The name of the write operation.
+     * @param tp    The address of the coordinator (aka its topic-partitions).
+     * @param op    The operation.
+     */
+    private void scheduleInternalOperation(
+        String name,
+        TopicPartition tp,
+        Runnable op
+    ) {
+        log.debug("Scheduled execution of internal operation {}.", name);
+        processor.enqueue(new InternalEvent(name, tp, op));
+    }
+
+    /**
+     * Builds a new coordinator state machine.
+     *
+     * @param context The coordinator context.
+     *
+     * @return The built coordinator state machine.
+     */
+    private S buildCoordinator(
+        CoordinatorContext context
+    ) {
+        return coordinatorBuilderSupplier
+            .get()
+            .withSnapshotRegistry(context.snapshotRegistry)
+            .build();
+    }
+
+    /**
+     * @return The topic partitions of the coordinators currently registered in the
+     * runtime.
+     */
+    public Set<TopicPartition> partitions() {
+        return new HashSet<>(coordinators.keySet());
+    }
+
+    /**
+     * Schedules the loading of a coordinator. This is called when the broker is elected as
+     * the leader for a partition.
+     *
+     * @param tp                The topic partition of the coordinator. Records from this
+     *                          partitions will be read and applied to the coordinator.
+     * @param partitionEpoch    The epoch of the partition.
```

Comment:
can we give a bit more description on what the difference is between this and context.epoch?

Reply:
There are basically the same: `context.epoch = partitionEpoch`. Is the name confusing?